The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9912036179/job/27386009252) [(plain)](https://github.com/rust-lang/rust/commit/2b2901a2f34fb55afd428d5c04a59aef1d752dcd/checks/27386009252/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.451   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.468 Collecting boolean-py==4.0
#13 0.474   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.489 Collecting chardet==5.1.0
---
#13 3.706 Building wheels for collected packages: reuse
#13 3.707   Building wheel for reuse (pyproject.toml): started
#13 4.047   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 4.048   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 4.049   Stored in directory: /tmp/pip-ephem-wheel-cache-bbb4ws_9/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 4.051 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.074   Attempting uninstall: setuptools
#13 4.075     Found existing installation: setuptools 59.6.0
#13 4.076     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.076     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.077     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.792 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.792 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.339 Collecting virtualenv
#13 5.385   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.590      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 28.2 MB/s eta 0:00:00
#13 5.632 Collecting distlib<1,>=0.3.7
#13 5.640   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.653      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 44.9 MB/s eta 0:00:00
#13 5.683 Collecting platformdirs<5,>=3.9.1
#13 5.690   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.726 Collecting filelock<4,>=3.12.2
#13 5.733   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.822 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 6.008 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 6.1s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      180160 kB
DirectMap2M:     9256960 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/5d76a13bbedebd773b4960432bff14f40acf3840/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-5d76a13bbedebd773b4960432bff14f40acf3840-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 29.01s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/std/src/io/buffered/tests.rs at line 1071:
 #[test]
 fn bufwriter_aliasing() {
-    use crate::io::{Cursor, BufWriter};
+    use crate::io::{BufWriter, Cursor};
+    use crate::io::{BufWriter, Cursor};
     let mut v = vec![0; 1024];
     let c = Cursor::new(&mut v);
     let w = BufWriter::new(Box::new(c));
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/io/prelude.rs" "/checkout/library/std/src/io/stdio/tests.rs" "/checkout/library/std/src/io/util/tests.rs" "/checkout/library/std/src/io/buffered/bufwriter.rs" "/checkout/library/std/src/io/buffered/bufreader/buffer.rs" "/checkout/library/std/src/io/buffered/bufreader.rs" "/checkout/library/std/src/io/buffered/linewriter.rs" "/checkout/library/std/src/io/buffered/linewritershim.rs" "/checkout/library/std/src/io/buffered/tests.rs" "/checkout/library/std/src/io/buffered/mod.rs" "/checkout/library/std/src/io/error/repr_unpacked.rs" "/checkout/library/std/src/io/error/repr_bitpacked.rs" "/checkout/library/std/src/io/error/tests.rs" "/checkout/library/std/src/io/util.rs" "/checkout/library/std/src/io/copy/tests.rs" "/checkout/library/std/src/io/tests.rs" "/checkout/library/std/src/io/mod.rs" "/checkout/library/std/src/time/tests.rs" "/checkout/library/std/src/env.rs" "/checkout/library/std/src/f16.rs" "/checkout/library/std/src/fs/tests.rs" "/checkout/library/std/build.rs" "/checkout/library/std/tests/eq-multidispatch.rs" "/checkout/library/std/tests/seq-compare.rs" "/checkout/library/std/tests/minmax-stability-issue-23687.rs" "/checkout/library/std/tests/process_spawning.rs" "/checkout/library/std/tests/create_dir_all_bare.rs" "/checkout/library/std/tests/type-name-unsized.rs" "/checkout/library/std/tests/builtin-clone.rs" "/checkout/library/std/tests/windows.rs" "/checkout/library/std/tests/run-time-detect.rs" "/checkout/library/std/tests/log-knows-the-names-of-variants-in-std.rs" "/checkout/library/std/tests/thread.rs" "/checkout/library/std/tests/switch-stdout.rs" "/checkout/library/std/tests/volatile-fat-ptr.rs" "/checkout/library/std/tests/istr.rs" "/checkout/library/std/tests/common/mod.rs" "/checkout/library/std/tests/slice-from-array-issue-113238.rs" "/checkout/library/std/tests/env.rs" "/checkout/library/std/benches/hash/set_ops.rs" "/checkout/library/std/benches/hash/map.rs" "/checkout/library/std/benches/hash/mod.rs" "/checkout/library/std/benches/lib.rs" "/checkout/library/rustc-std-workspace-core/lib.rs" "/checkout/library/rtstartup/rsbegin.rs" "/checkout/library/rtstartup/rsend.rs" "/checkout/library/rustc-std-workspace-std/lib.rs" "/checkout/library/panic_abort/src/android.rs" "/checkout/library/panic_abort/src/lib.rs" "/checkout/library/panic_abort/src/zkvm.rs" "/checkout/library/rustc-std-workspace-alloc/lib.rs" "/checkout/library/alloc/src/ffi/c_str/tests.rs" "/checkout/library/alloc/src/ffi/mod.rs" "/checkout/library/alloc/src/ffi/c_str.rs" "/checkout/library/alloc/src/borrow.rs" "/checkout/library/alloc/src/rc.rs" "/checkout/library/alloc/src/macros.rs" "/checkout/library/alloc/src/collections/linked_list.rs" "/checkout/library/alloc/src/collections/vec_deque/macros.rs" "/checkout/library/alloc/src/collections/vec_deque/iter.rs" "/checkout/library/alloc/src/collections/vec_deque/spec_from_iter.rs" "/checkout/library/alloc/src/collections/vec_deque/into_iter.rs" "/checkout/library/alloc/src/collections/vec_deque/spec_extend.rs" "/checkout/library/std/src/io/impls.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Jul 12 17:56:45 UTC 2024
  network time: Fri, 12 Jul 2024 17:56:45 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/library/std/src/io/buffered/bufwriter.rs
+++ b/library/std/src/io/buffered/bufwriter.rs
@@ -3,7 +3,7 @@ use crate::fmt;
 use crate::io::{
     self, ErrorKind, IntoInnerError, IoSlice, Seek, SeekFrom, Write, DEFAULT_BUF_SIZE,
 };
-use crate::mem;
+use crate::mem::{self, ManuallyDrop};
 use crate::ptr;
 
 /// Wraps a writer and buffers its output.
@@ -164,13 +164,13 @@ impl<W: Write> BufWriter<W> {
     /// assert_eq!(&buffered_data.unwrap(), b"ata");
     /// ```
     #[stable(feature = "bufwriter_into_parts", since = "1.56.0")]
-    pub fn into_parts(mut self) -> (W, Result<Vec<u8>, WriterPanicked>) {
-        let buf = mem::take(&mut self.buf);
-        let buf = if !self.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+    pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+        let mut this = ManuallyDrop::new(self);
+        let buf = mem::take(&mut this.buf);
+        let buf = if !this.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
 
-        // SAFETY: forget(self) prevents double dropping inner
-        let inner = unsafe { ptr::read(&self.inner) };
-        mem::forget(self);
+        // SAFETY: double-drops are prevented by putting `this` in a ManuallyDrop that is never dropped
+        let inner = unsafe { ptr::read(&this.inner) };
 
         (inner, buf)
     }

--- a/library/std/src/io/buffered/tests.rs
+++ b/library/std/src/io/buffered/tests.rs
@@ -1067,3 +1067,13 @@ fn bufreader_full_initialize() {
     // But we initialized the whole buffer!
     assert_eq!(reader.initialized(), reader.capacity());
 }
+
+/// This is a regression test for https://github.com/rust-lang/rust/issues/127584.
+#[test]
+fn bufwriter_aliasing() {
+    use crate::io::{BufWriter, Cursor};
+    let mut v = vec![0; 1024];
+    let c = Cursor::new(&mut v);
+    let w = BufWriter::new(Box::new(c));
+    let _ = w.into_parts();
+}

